hta/common/trace_parser.py

```diff
@@ -428,3 +436,102 @@ def parse_trace_dataframe(
+    This leads to some blazingly fast metadata parsing!
+
+    https://pypi.org/project/ijson/#toc-entry-4
+
+    The ijson low-level API has a series of (prefix, event, value). The
+    prefix shows the path in the json object, while event can signify
+    the start of a map, array or a key. Following is an excerpt of events for a trace file
+
+    ```
+     start_map None
+     map_key schemaVersion
+    schemaVersion number 1
+     map_key distributedInfo
+    distributedInfo start_map None
+    distributedInfo map_key backend
+    distributedInfo.backend string nccl
+    distributedInfo map_key rank
+    distributedInfo.rank number 0
+    distributedInfo map_key world_size
+    distributedInfo.world_size number 64
+    distributedInfo end_map None
+     map_key deviceProperties
+    deviceProperties start_array None
+    deviceProperties.item start_map None
+    deviceProperties.item map_key id
+    deviceProperties.item.id number 0
+    ```
+    """
+    import ijson
+
+    meta: MetaData = {}
+    trace_parser = ijson.parse(fh)
+
+    def handle_nested_map(iterator, key: str, meta_so_far: MetaData) -> MetaData:
+        """
+        Recursively handles events for a nested map like distributedInfo.
+        """
+        prefix, event, value = next(iterator)
+        logger.debug(" -> ", prefix, event, value)
+        if event == "end_map":
+            return meta_so_far
+        elif event == "map_key":
+            key = value
+            return handle_nested_map(iterator, key, meta_so_far)
+        else:
+            assert key is not None
+            meta_so_far[key] = value
+            return handle_nested_map(iterator, "", meta_so_far)
+
+    cur_key = None
+    nested_map: MetaData = {}
+
+    for prefix, event, value in trace_parser:
+        logger.debug(prefix, event, value)
+        if event == "map_key" and value == "traceEvents":
+            # done ok bye!
+            return meta
+
+        # Handle a nested map like "distributedInfo"
+        if prefix == cur_key and event == "start_map":
+            nested_map = {}
+            meta[cur_key] = handle_nested_map(trace_parser, "", nested_map)
+            continue
+
+        # Handle a nested array map like "deviceProperties"
+        if prefix == cur_key and event == "start_array":
+            meta[cur_key] = []
+            # this should be start_map
+            _, _event_, _ = next(trace_parser)
+            assert _event_ == "start_map"
+            while _event_ != "end_array":
+                nested_map = {}
+                nested_map = handle_nested_map(trace_parser, "", nested_map)
+                meta[cur_key].append(nested_map)
+                _, _event_, _ = next(trace_parser)
+            continue
+
+        # Handle top level simple key values
+        #  map_key schemaVersion
+        # schemaVersion number 1
+        if event == "map_key" and prefix == "":
+            cur_key = value
```

Comment:
seems this is the only place `cur_key` is updated, but it's used as the key in `meta` in many places above. is it possible to hit one of the `meta[cur_key]` above when `cur_key` is `None`?

Reply:
Ideally its not possible as there should always be a map_key event before any value. can add an assert